src/tools/illink/src/linker/Linker/TypeMapInfo.cs

```diff
@@ -158,36 +157,35 @@ static void AddRecursiveInterfaces (TypeReference typeRef, IEnumerable<Interface
 					if (interfaceType is null) {
 						continue;
 					}
-					if (!firstImplementationChain.Any (i => TypeReferenceEqualityComparer.AreEqual (i.Item1, interfaceType, Context))) {
-						firstImplementationChain.Add ((interfaceType, pathToType.Append (iface).ToList ()));
+					if (!implementors.Any (i => TypeReferenceEqualityComparer.AreEqual (i.InflatedInterfaceType, interfaceType, Context))) {
+						implementors.Add (new RuntimeInterfaceImplementation (originalType, typeRef, pathToType.Append (iface), interfaceType, Context));
 					}
 				}
 
+				if (type.BaseType is not null) {
+					var inflatedBaseType = type.BaseType.TryInflateFrom (typeRef, Context)!;
+					AddRecursiveInterfaces (originalType, inflatedBaseType, [], implementors, Context);
```

Comment:
I was going off the implements/inherits tree and type declaration order in the ecma spec, which I think describes this ordering, though I might be misunderstanding. It looks like NativeAOT does it this way, though: https://github.com/dotnet/runtime/blob/ac996f3f48ea8c3364c347c24962a75fd80eafae/src/coreclr/tools/Common/TypeSystem/Common/MetadataRuntimeInterfacesAlgorithm.cs#L49-L86

Reply:
The latest changes make it so that the order doesn't matter. It will look at all recursive interfaces and add their implementations as DIMs regardless of where it lies in the inherits/implements order.

Whenever an interface/implementor relationship (a RuntimeInterfaceImplementation) is marked, it will mark all the possible interfaceImpl chains that can cause the "implementor" to implement the interface at runtime. So, an explicit interfaceImpl will be marked, but also every recursive chain that ends with the interface type.